include/deal.II/lac/tensor_product_matrix.h

```diff
@@ -512,61 +674,17 @@ TensorProductMatrixSymmetricSumBase<dim, Number, n_rows_1d>::vmult(
   AssertDimension(dst_view.size(), this->m());
   AssertDimension(src_view.size(), this->n());
   std::lock_guard<std::mutex> lock(this->mutex);
-  const unsigned int          n = Utilities::fixed_power<dim>(
-    n_rows_1d > 0 ? n_rows_1d : eigenvalues[0].size());
-  tmp_array.resize_fast(n * 2);
-  constexpr int kernel_size = n_rows_1d > 0 ? n_rows_1d : 0;
-  internal::EvaluatorTensorProduct<internal::evaluate_general,
-                                   dim,
-                                   kernel_size,
-                                   kernel_size,
-                                   Number>
-                eval(AlignedVector<Number>{},
-         AlignedVector<Number>{},
-         AlignedVector<Number>{},
-         mass_matrix[0].n_rows(),
-         mass_matrix[0].n_rows());
-  Number *      t   = tmp_array.begin();
-  const Number *src = src_view.begin();
-  Number *      dst = dst_view.data();
 
-  if (dim == 1)
-    {
-      const Number *A = &derivative_matrix[0](0, 0);
-      eval.template apply<0, false, false>(A, src, dst);
-    }
-
-  else if (dim == 2)
-    {
-      const Number *A0 = &derivative_matrix[0](0, 0);
-      const Number *M0 = &mass_matrix[0](0, 0);
-      const Number *A1 = &derivative_matrix[1](0, 0);
-      const Number *M1 = &mass_matrix[1](0, 0);
-      eval.template apply<0, false, false>(M0, src, t);
-      eval.template apply<1, false, false>(A1, t, dst);
-      eval.template apply<0, false, false>(A0, src, t);
-      eval.template apply<1, false, true>(M1, t, dst);
-    }
-
-  else if (dim == 3)
-    {
-      const Number *A0 = &derivative_matrix[0](0, 0);
-      const Number *M0 = &mass_matrix[0](0, 0);
-      const Number *A1 = &derivative_matrix[1](0, 0);
-      const Number *M1 = &mass_matrix[1](0, 0);
-      const Number *A2 = &derivative_matrix[2](0, 0);
-      const Number *M2 = &mass_matrix[2](0, 0);
-      eval.template apply<0, false, false>(M0, src, t + n);
-      eval.template apply<1, false, false>(M1, t + n, t);
-      eval.template apply<2, false, false>(A2, t, dst);
-      eval.template apply<1, false, false>(A1, t + n, t);
-      eval.template apply<0, false, false>(A0, src, t + n);
-      eval.template apply<1, false, true>(M1, t + n, t);
-      eval.template apply<2, false, true>(M2, t, dst);
-    }
+  Number *      dst = dst_view.begin();
+  const Number *src = src_view.begin();
 
+  if (n_rows_1d != -1)
+    internal::TensorProductMatrixSymmetricSum::vmult<
+      n_rows_1d == -1 ? 0 : n_rows_1d>(
```

Comment:
It looks like you will never reach the `n_rows_1d == -1` in this branch, as you check `n_rows_1d != -1` in line 681. Is the code there to reconcile the compiler?

Reply:
The line is never reached; but without `if consexpr`, the compiler will still complain.

Reply:
I know, I've had similar constructions before.